src/transaction/validation.rs

```diff
@@ -134,19 +150,35 @@ impl Input {
 }
 
 impl Output {
+    /// Validate the output of the transaction.
+    ///
+    /// This function is stateful - meaning it might validate a transaction during VM
+    /// initialization, but this transaction will no longer be valid in post-execution because the
+    /// VM might mutate the message outputs, producing invalid transactions.
```

Comment:
> producing invalid transactions.

If the VM is implemented correctly (we can assume it is), then a valid transaction that's input into the VM will output a valid transaction. Anything else would be a DoS vector.

Reply:
We do override the recipient and amount of the output:

https://github.com/FuelLabs/fuel-specs/blob/e0d5d8aea3f6c344ca47776403accc58b67eeb01/specs/vm/instruction_set.md?plain=1#L1417-L1420

It sounds odd that we execute a transaction (with zeroed recipient/amount) different than the one that was signed (with original recipient/amount). If a transaction is submitted with these fields filled, shouldn't it be invalid?

Reply:
It should be invalid, yes, but why is this stateful is my confusion.

Reply:
This is ofc a matter of perspective, but that would be the "pre-execution" and "post-execution" states of a transaction. I will alter the behavior to produce valid transactions for both conditions and just force these values to zero during VM initialization (as it is literally described in the specs).

I just feel a bit uncomfortable of mutating the tx during VM initialization since it creates a magic box effect

Reply:
https://github.com/FuelLabs/fuel-tx/pull/149/commits/411a630c2ecacba80711c038d65e03afa1506262